File: swift_reference/Views/Onboarding/PaywallScreen.swift
import SwiftUI
import RevenueCat

struct PaywallScreen: View {
    let store: StoreViewModel
    let onContinue: () -> Void

    @State private var showContent: Bool = false
    @State private var showFeatures: Bool = false
    @State private var selectedPlan: PlanType = .yearly

    private enum PlanType {
        case monthly, yearly
    }

    private let freeFeatures: [(icon: String, text: String)] = [
        ("hands.sparkles.fill", "Daily Gratitude"),
        ("plus.circle", "2 Custom Habits"),
        ("calendar", "Weekly View"),
        ("book.fill", "Anchor Verses"),
    ]

    private let proFeatures: [(icon: String, text: String)] = [
        ("infinity", "Unlimited Habits"),
        ("text.quote", "Custom Purpose Statements"),
        ("chart.bar.fill", "Detailed Stats & Insights"),
        ("bell.badge.fill", "Smart Reminders"),
        ("shield.fill", "SOS Temptation Support"),
        ("sparkles", "52-Week Year in Tribute"),
    ]

    var body: some View {
        VStack(spacing: 0) {
            ScrollView {
                VStack(spacing: 28) {
                    VStack(spacing: 10) {
                        Text("Go deeper with\nTribute Pro")
                            .font(.system(.title2, design: .serif, weight: .bold))
                            .foregroundStyle(.primary)
                            .multilineTextAlignment(.center)

                        Text("Everything you need to build lasting habits rooted in faith.")
                            .font(.subheadline)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                    }
                    .opacity(showContent ? 1 : 0)

                    HStack(alignment: .top, spacing: 12) {
                        VStack(alignment: .leading, spacing: 14) {
                            Text("FREE")
                                .font(.caption.weight(.bold))
                                .foregroundStyle(.secondary)
                                .tracking(1.5)

                            ForEach(freeFeatures, id: \.text) { feature in
                                HStack(spacing: 10) {
                                    Image(systemName: feature.icon)
                                        .font(.system(size: 13))
                                        .foregroundStyle(TributeColor.softGold.opacity(0.5))
                                        .frame(width: 18)

                                    Text(feature.text)
                                        .font(.caption)
                                        .foregroundStyle(.secondary)
                                }
                            }
                        }
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(16)
                        .background(TributeColor.cardBackground)
                        .clipShape(.rect(cornerRadius: 14))
                        .overlay(
                            RoundedRectangle(cornerRadius: 14)
                                .strokeBorder(TributeColor.cardBorder, lineWidth: 0.5)
                        )

                        VStack(alignment: .leading, spacing: 14) {
                            HStack(spacing: 6) {
                                Text("PRO")
                                    .font(.caption.weight(.bold))
                                    .foregroundStyle(TributeColor.golden)
                                    .tracking(1.5)

                                Image(systemName: "crown.fill")
                                    .font(.system(size: 10))
                                    .foregroundStyle(TributeColor.golden)
                            }

                            ForEach(proFeatures, id: \.text) { feature in
                                HStack(spacing: 10) {
                                    Image(systemName: feature.icon)
                                        .font(.system(size: 13))
                                        .foregroundStyle(TributeColor.golden)
                                        .frame(width: 18)

                                    Text(feature.text)
                                        .font(.caption)
                                        .foregroundStyle(.primary)
                                }
                            }
                        }
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(16)
                        .background(TributeColor.golden.opacity(0.06))
                        .clipShape(.rect(cornerRadius: 14))
                        .overlay(
                            RoundedRectangle(cornerRadius: 14)
                                .strokeBorder(TributeColor.golden.opacity(0.2), lineWidth: 0.5)
                        )
                    }
                    .opacity(showFeatures ? 1 : 0)
                    .offset(y: showFeatures ? 0 : 12)

                    planSelector
                        .opacity(showFeatures ? 1 : 0)
                }
                .padding(.horizontal, 24)
                .padding(.top, 16)
                .padding(.bottom, 32)
            }

            VStack(spacing: 12) {
                Button {
                    Task {
                        let pkg = selectedPlan == .yearly ? store.annualPackage : store.monthlyPackage
                        if let pkg {
                            await store.purchase(package: pkg)
                            if store.isPremium {
                                onContinue()
                            }
                        } else {
                            onContinue()
                        }
                    }
                } label: {
                    HStack(spacing: 8) {
                        if store.isPurchasing {
                            ProgressView()
                                .tint(TributeColor.charcoal)
                        } else {
                            Image(systemName: "crown.fill")
                                .font(.subheadline)
                            Text("Start Free Trial")
                        }
                    }
                    .tributeButton()
                }
                .disabled(store.isPurchasing)

                HStack(spacing: 16) {
                    Button {
                        Task { await store.restore() }
                    } label: {
                        Text("Restore")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }

                    Button {
                        onContinue()
                    } label: {
                        Text("Continue with Free")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 32)
        }
        .onAppear {
            withAnimation(.easeOut(duration: 0.5).delay(0.2)) {
                showContent = true
            }
            withAnimation(.easeOut(duration: 0.5).delay(0.5)) {
                showFeatures = true
            }
        }
    }

    private var planSelector: some View {
        HStack(spacing: 12) {
            if let monthly = store.monthlyPackage {
                planOption(
                    title: "Monthly",
                    price: monthly.storeProduct.localizedPriceString,
                    detail: "per month",
                    isSelected: selectedPlan == .monthly,
                    badge: nil
                ) {
                    selectedPlan = .monthly
                }
            }

            if let annual = store.annualPackage {
                let monthlyEquiv = annual.storeProduct.price / 12
                let formatter = NumberFormatter()
                let _ = { formatter.numberStyle = .currency; formatter.locale = annual.storeProduct.priceFormatter?.locale ?? .current }()
                let monthlyStr = formatter.string(from: monthlyEquiv as NSDecimalNumber) ?? ""

                planOption(
                    title: "Yearly",
                    price: annual.storeProduct.localizedPriceString,
                    detail: "\(monthlyStr)/mo",
                    isSelected: selectedPlan == .yearly,
                    badge: store.monthlySavingsText
                ) {
                    selectedPlan = .yearly
                }
            }
        }
    }

    private func planOption(
        title: String,
        price: String,
        detail: String,
        isSelected: Bool,
        badge: String?,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            VStack(spacing: 6) {
                if let badge {
                    Text(badge)
                        .font(.system(size: 9, weight: .bold))
                        .foregroundStyle(TributeColor.charcoal)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(TributeColor.golden)
                        .clipShape(Capsule())
                } else {
                    Spacer().frame(height: 17)
                }

                Text(title)
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(isSelected ? TributeColor.golden : .secondary)

                Text(price)
                    .font(.system(.subheadline, design: .rounded, weight: .bold))
                    .foregroundStyle(isSelected ? .primary : .secondary)

                Text(detail)
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 14)
            .background(isSelected ? TributeColor.golden.opacity(0.08) : TributeColor.cardBackground)
            .clipShape(.rect(cornerRadius: 12))
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .strokeBorder(
                        isSelected ? TributeColor.golden.opacity(0.4) : TributeColor.cardBorder,
                        lineWidth: isSelected ? 1.5 : 0.5
                    )
            )
        }
        .buttonStyle(.plain)
    }
}

File: lib/presentation/views/onboarding/paywall_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../providers/store_provider.dart';
import '../../theme/app_theme.dart';

enum _Plan { monthly, annual, lifetime }

class PaywallScreen extends StatefulWidget {
  final VoidCallback onNext;
  const PaywallScreen({super.key, required this.onNext});

  @override
  State<PaywallScreen> createState() => _PaywallScreenState();
}

class _PaywallScreenState extends State<PaywallScreen> {
  _Plan _selectedPlan = _Plan.annual;
  bool _showContent = false;
  bool _showFeatures = false;
  StoreProvider? _store;

  static const _freeFeatures = [
    (Icons.volunteer_activism, 'Daily Gratitude'),
    (Icons.add_circle_outline, '2 Custom Habits'),
    (Icons.calendar_today, 'Weekly View'),
    (Icons.menu_book, 'Anchor Verses'),
  ];

  static const _proFeatures = [
    (Icons.all_inclusive_rounded, 'Unlimited Habits'),
    (Icons.format_quote_rounded, 'Custom Purpose Statements'),
    (Icons.bar_chart_rounded, 'Detailed Stats & Insights'),
    (Icons.notifications_rounded, 'Smart Reminders'),
    (Icons.shield_rounded, 'SOS Temptation Support'),
    (Icons.auto_awesome, '52-Week Year in MyWalk'),
  ];

  @override
  void initState() {
    super.initState();
    Future.delayed(const Duration(milliseconds: 200), () {
      if (mounted) setState(() => _showContent = true);
    });
    Future.delayed(const Duration(milliseconds: 500), () {
      if (mounted) setState(() => _showFeatures = true);
    });
    WidgetsBinding.instance.addPostFrameCallback((_) {
      if (!mounted) return;
      _store = context.read<StoreProvider>();
      _store!.addListener(_onStoreChanged);
    });
  }

  @override
  void dispose() {
    _store?.removeListener(_onStoreChanged);
    super.dispose();
  }

  void _onStoreChanged() {
    final store = _store;
    if (store == null || !mounted) return;
    // Fire onNext when premium is gained via either a new purchase or a restore.
    if (store.isPremium && !store.isPurchasing) {
      widget.onNext();
    }
  }

  String get _ctaLabel => switch (_selectedPlan) {
        _Plan.monthly => 'Subscribe Monthly',
        _Plan.annual => 'Start Free Trial',
        _Plan.lifetime => 'Buy Lifetime Access',
      };

  Future<void> _purchase(StoreProvider store) async {
    final product = switch (_selectedPlan) {
      _Plan.monthly => store.monthlyProduct,
      _Plan.annual => store.annualProduct,
      _Plan.lifetime => store.lifetimeProduct,
    };
    if (product != null) {
      await store.purchase(product);
      // Navigation is handled by _onStoreChanged when isPremium becomes true.
    } else {
      // Product unavailable (store unreachable) — let user continue on free.
      widget.onNext();
    }
  }

  @override
  Widget build(BuildContext context) {
    final store = context.watch<StoreProvider>();
    return Column(children: [
      Expanded(
        child: SingleChildScrollView(
          padding: const EdgeInsets.fromLTRB(24, 16, 24, 32),
          child: Column(children: [
            AnimatedOpacity(
              opacity: _showContent ? 1.0 : 0.0,
              duration: const Duration(milliseconds: 500),
              child: Column(children: [
                const Text(
                  'Go deeper with\nMyWalk Pro',
                  textAlign: TextAlign.center,
                  style: TextStyle(
                      fontSize: 22,
                      fontWeight: FontWeight.w700,
                      color: MyWalkColor.warmWhite,
                      height: 1.3),
                ),
                const SizedBox(height: 10),
                Text(
                  'Everything you need to build lasting habits rooted in faith.',
                  textAlign: TextAlign.center,
                  style: TextStyle(
                      fontSize: 15, color: Colors.white.withValues(alpha: 0.5)),
                ),
              ]),
            ),
            const SizedBox(height: 28),
            AnimatedOpacity(
              opacity: _showFeatures ? 1.0 : 0.0,
              duration: const Duration(milliseconds: 500),
              child: AnimatedSlide(
                offset: _showFeatures ? Offset.zero : const Offset(0, 0.15),
                duration: const Duration(milliseconds: 500),
                child: Row(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Expanded(
                          child: _featureColumn(
                              label: 'FREE',
                              isGold: false,
                              features: _freeFeatures)),
                      const SizedBox(width: 12),
                      Expanded(
                          child: _featureColumn(
                              label: 'PRO',
                              isGold: true,
                              features: _proFeatures)),
                    ]),
              ),
            ),
            const SizedBox(height: 20),
            AnimatedOpacity(
              opacity: _showFeatures ? 1.0 : 0.0,
              duration: const Duration(milliseconds: 500),
              child: _planSelector(store),
            ),
          ]),
        ),
      ),
      Padding(
        padding: const EdgeInsets.fromLTRB(24, 0, 24, 32),
        child: Column(children: [
          SizedBox(
            width: double.infinity,
            child: ElevatedButton.icon(
              onPressed: store.isPurchasing ? null : () => _purchase(store),
              icon: store.isPurchasing
                  ? const SizedBox(
                      width: 18,
                      height: 18,
                      child: CircularProgressIndicator(
                          strokeWidth: 2, color: MyWalkColor.charcoal))
                  : const Icon(Icons.workspace_premium_rounded, size: 18),
              label: Text(_ctaLabel,
                  style: const TextStyle(
                      fontWeight: FontWeight.w600, fontSize: 16)),
              style: ElevatedButton.styleFrom(
                backgroundColor: MyWalkColor.golden,
                foregroundColor: MyWalkColor.charcoal,
                padding: const EdgeInsets.symmetric(vertical: 16),
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(14)),
              ),
            ),
          ),
          const SizedBox(height: 12),
          Row(mainAxisAlignment: MainAxisAlignment.center, children: [
            TextButton(
              onPressed: store.isLoading ? null : () => store.restore(),
              child: Text('Restore',
                  style: TextStyle(
                      fontSize: 12,
                      color: Colors.white.withValues(alpha: 0.5))),
            ),
            const SizedBox(width: 16),
            TextButton(
              onPressed: widget.onNext,
              child: Text('Continue with Free',
                  style: TextStyle(
                      fontSize: 12,
                      color: Colors.white.withValues(alpha: 0.5))),
            ),
          ]),
          if (store.error != null)
            Padding(
              padding: const EdgeInsets.only(top: 8),
              child: Text(store.error!,
                  textAlign: TextAlign.center,
                  style: const TextStyle(
                      fontSize: 12, color: MyWalkColor.warmCoral)),
            ),
        ]),
      ),
    ]);
  }

  Widget _featureColumn({
    required String label,
    required bool isGold,
    required List<(IconData, String)> features,
  }) {
    return Container(
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: isGold
            ? MyWalkColor.golden.withValues(alpha: 0.06)
            : MyWalkColor.cardBackground,
        borderRadius: BorderRadius.circular(14),
        border: Border.all(
          color: isGold
              ? MyWalkColor.golden.withValues(alpha: 0.2)
              : MyWalkColor.cardBorder,
          width: 0.5,
        ),
      ),
      child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
        Row(children: [
          Text(label,
              style: TextStyle(
                fontSize: 11,
                fontWeight: FontWeight.w700,
                letterSpacing: 1.5,
                color: isGold
                    ? MyWalkColor.golden
                    : Colors.white.withValues(alpha: 0.5),
              )),
          if (isGold) ...[
            const SizedBox(width: 6),
            const Icon(Icons.workspace_premium_rounded,
                size: 10, color: MyWalkColor.golden),
          ],
        ]),
        const SizedBox(height: 14),
        ...features.map((f) => Padding(
              padding: const EdgeInsets.only(bottom: 12),
              child: Row(children: [
                SizedBox(
                  width: 18,
                  child: Icon(f.$1,
                      size: 13,
                      color: isGold
                          ? MyWalkColor.golden
                          : MyWalkColor.softGold.withValues(alpha: 0.5)),
                ),
                const SizedBox(width: 10),
                Expanded(
                  child: Text(f.$2,
                      style: TextStyle(
                        fontSize: 12,
                        color: isGold
                            ? MyWalkColor.warmWhite
                            : Colors.white.withValues(alpha: 0.5),
                      )),
                ),
              ]),
            )),
      ]),
    );
  }

  Widget _planSelector(StoreProvider store) {
    final monthly = store.monthlyProduct;
    final annual = store.annualProduct;
    final lifetime = store.lifetimeProduct;

    if (monthly == null && annual == null && lifetime == null) {
      return Text('Loading plans\u2026',
          style:
              TextStyle(fontSize: 13, color: Colors.white.withValues(alpha: 0.4)));
    }

    return Column(children: [
      // Subscription row: monthly + annual side by side.
      if (monthly != null || annual != null)
        Row(children: [
          if (monthly != null) ...[
            Expanded(
                child: _planOption(
              title: 'Monthly',
              price: monthly.price,
              detail: 'per month',
              trialText: null,
              isSelected: _selectedPlan == _Plan.monthly,
              badge: null,
              onTap: () => setState(() => _selectedPlan = _Plan.monthly),
            )),
            const SizedBox(width: 12),
          ],
          if (annual != null)
            Expanded(
                child: _planOption(
              title: 'Yearly',
              price: annual.price,
              detail: '\$${(annual.rawPrice / 365).toStringAsFixed(2)}/day',
              trialText: '7-day free trial',
              isSelected: _selectedPlan == _Plan.annual,
              badge: store.monthlySavingsText,
              onTap: () => setState(() => _selectedPlan = _Plan.annual),
            )),
        ]),
      // Lifetime: full-width card below.
      if (lifetime != null) ...[
        if (monthly != null || annual != null) const SizedBox(height: 12),
        _planOption(
          title: 'Lifetime',
          price: lifetime.price,
          detail: 'one-time purchase · never expires',
          trialText: null,
          isSelected: _selectedPlan == _Plan.lifetime,
          badge: 'Best Deal',
          onTap: () => setState(() => _selectedPlan = _Plan.lifetime),
        ),
      ],
    ]);
  }

  Widget _planOption({
    required String title,
    required String price,
    required String detail,
    required String? trialText,
    required bool isSelected,
    required String? badge,
    required VoidCallback onTap,
  }) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(vertical: 14),
        decoration: BoxDecoration(
          color: isSelected
              ? MyWalkColor.golden.withValues(alpha: 0.08)
              : MyWalkColor.cardBackground,
          borderRadius: BorderRadius.circular(12),
          border: Border.all(
            color: isSelected
                ? MyWalkColor.golden.withValues(alpha: 0.4)
                : MyWalkColor.cardBorder,
            width: isSelected ? 1.5 : 0.5,
          ),
        ),
        child: Column(children: [
          if (badge != null)
            Container(
              padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
              decoration: BoxDecoration(
                  color: MyWalkColor.golden,
                  borderRadius: BorderRadius.circular(20)),
              child: Text(badge,
                  style: const TextStyle(
                      fontSize: 9,
                      fontWeight: FontWeight.w700,
                      color: MyWalkColor.charcoal)),
            )
          else
            const SizedBox(height: 17),
          const SizedBox(height: 4),
          Text(title,
              style: TextStyle(
                fontSize: 11,
                fontWeight: FontWeight.w600,
                letterSpacing: 0.5,
                color: isSelected
                    ? MyWalkColor.golden
                    : Colors.white.withValues(alpha: 0.5),
              )),
          const SizedBox(height: 4),
          Text(price,
              style: TextStyle(
                fontSize: 15,
                fontWeight: FontWeight.w700,
                color: isSelected
                    ? MyWalkColor.warmWhite
                    : Colors.white.withValues(alpha: 0.5),
              )),
          const SizedBox(height: 2),
          Text(detail,
              style:
                  TextStyle(fontSize: 11, color: Colors.white.withValues(alpha: 0.4))),
          if (trialText != null) ...[
            const SizedBox(height: 4),
            Text(trialText,
                style: TextStyle(
                    fontSize: 9,
                    fontWeight: FontWeight.w600,
                    color: isSelected
                        ? MyWalkColor.sage
                        : Colors.white.withValues(alpha: 0.3))),
          ],
        ]),
      ),
    );
  }
}
